docs/design/phases.md

```diff
@@ -0,0 +1,354 @@
+ - `kubeadm phase controlplane apiserver`
+   - Inputs:
+     - `--image=gcr.io/google_containers/kube-apiserver-${ARCH}:${STABLE_VERSION}`
+     - `--service-cidr=10.96.0.1/12`
+     - `--cert-dir=/etc/kubernetes/pki`
+     - `--port=6443`
+     - `--address=0.0.0.0`
+     - `--etcd-servers=http://localhost:2379`
+     - `--authorization-mode=RBAC`
+     - `--extra-args`
+     - `--extra-volumes`
+   - Output:
+     - `kube-apiserver.yaml` in the manifests dir (`/etc/kubernetes/manifests`)
+ - `kubeadm phase controlplane controller-manager`
+   - Inputs:
+     - `--image=gcr.io/google_containers/kube-controller-manager-${ARCH}:${STABLE_VERSION}`
+     - `--kubeconfig=/etc/kubernetes/controller-manager.conf`
+     - `--pod-network-cidr`
+     - `--cert-dir=/etc/kubernetes/pki`
+     - `--auto-approve-csrs=true`
+     - `--extra-args`
+     - `--extra-volumes`
+   - Output:
+     - `kube-controller-manager.yaml` in the manifests dir (`/etc/kubernetes/manifests`)
+ - `kubeadm phase controlplane scheduler`
+   - Inputs:
+     - `--image=gcr.io/google_containers/kube-scheduler-${ARCH}:${STABLE_VERSION}`
+     - `--kubeconfig=/etc/kubernetes/scheduler.conf`
+     - `--extra-args`
+   - Output:
+     - `kube-scheduler.yaml` in the manifests dir (`/etc/kubernetes/manifests`)
+ - `kubeadm phase addons cluster-info`
+   - Inputs:
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+   - Outputs:
+     - Creates the `cluster-info` ConfigMap in the `kube-public` namespace
+     - Exposes it publicly with RBAC bindings
+- `kubeadm phase addons self-hosted`
+   - Inputs:
+     - Assumes the control plane yaml manifests are in `/etc/kubernetes/manifests`
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+     - `--cert-dir=/etc/kubernetes/pki`
+   - Outputs:
+     - Creates a selfhosted control plane from static pod manifests
+- `kubeadm phase addons kube-proxy`
+   - Inputs:
+     - `--image=gcr.io/google_containers/kube-proxy-${ARCH}:${STABLE_VERSION}`
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+     - `--pod-network-cidr=`
+   - Outputs:
+     - Applies the built-in `kube-proxy` manifest to the API Server
+- `kubeadm phase addons kube-dns`
+   - Inputs:
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+     - `--dns-domain=cluster.local`
+     - `--image-repository=gcr.io/google_containers`
+   - Outputs:
+     - Applies the built-in `kube-dns` manifest to the API Server
+
+### `kubeadm init` UX and function in the light of `kubeadm phase`
+
+ - `kubeadm init`
+   - `--cert-dir=/etc/kubernetes/pki`
+   - `--service-cidr=10.96.0.1/12`
+   - `--cert-altnames=""`
+   - `--apiserver-address=0.0.0.0`
+   - `--apiserver-port=6443`
+   - `--image-repository=gcr.io/google_containers`
+   - `--pod-network-cidr=`
+   - `--dns-domain=cluster.local`
+   - `--token=${AUTOGENERATE}`
+   - `--token-ttl=0`
+   - `--use-kubernetes-version=${STABLE_VERSION}`
+   - `--self-hosted`
+
+What would `kubeadm init` actually do?
```

Comment:
Nice to have: Adding Aliases for each phase command with a numbered prefix could help users in understanding/using the right sequence of phases. e.g. 
> `kubeadm phase 1-certs ..` (where `1-certs` is alias of `certs`certificates`)
> `kubeadm phase 2-kubeconfig ..` (where `2-kubeconfig` is alias of `kubeconfig`)
Nb. this feature suites very well with autocompletion